airflow/jobs/triggerer_job_runner.py

```diff
@@ -643,7 +648,9 @@ async def run_trigger(self, trigger_id, trigger):
             # allow triggers a chance to cleanup, either in that case or if
             # they exit cleanly. Exception from cleanup methods are ignored.
             with suppress(Exception):
-                await trigger.cleanup()
+                trigger_details = self.triggers[trigger_id]
+                if trigger.should_cleanup(trigger_details["termination_reason"]):
```

Comment:
One reason I added the new method is that I didn’t want to modify the existing cleanup interface, which allows for backward compatibility with user code (please correct me if I’m wrong). The downside is that it’s not as simple and direct as you mentioned.

Reply:
And that is a very reasonable concern @TakawaAkirayo.  However right now with airflow 3 coming, we have a special opportunity where we can break backward compatibility.  Even in 2.x we could have done it by adding some kind of compatibility layer / conditional handling.  But since this will only make it into 3.x we don't need that.  What we will need to do though, is update providers to be forward-compatible with this change, so that new provider releases will be compat with 2.x and 3.x as much as possible.  E.g. maybe changing existing trigger cleanup signatures to take **kwargs.

Reply:
Got it. I've already removed the should_cleanup and added the termination context to the cleanup method. The documentation has been updated as well. @dstandish